indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java

```diff
@@ -56,38 +56,29 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
 
   public Optional<SupervisorSpec> getSupervisorSpec(String id)
   {
-    return supervisors.get(id) == null
-           ? Optional.<SupervisorSpec>absent()
-           : Optional.fromNullable(supervisors.get(id).rhs);
+    Pair<Supervisor, SupervisorSpec> supervisor = supervisors.get(id);
+    return supervisor == null ? Optional.<SupervisorSpec>absent() : Optional.fromNullable(supervisor.rhs);
   }
 
-  public boolean hasSupervisor(String id)
-  {
-    return supervisors.containsKey(id);
-  }
-
-  public boolean createAndStartSupervisor(SupervisorSpec spec)
+  public synchronized boolean createOrUpdateAndStartSupervisor(SupervisorSpec spec)
   {
     Preconditions.checkNotNull(spec, "spec");
     Preconditions.checkNotNull(spec.getId(), "spec.getId()");
 
+    possiblyStopAndRemoveSupervisorInternal(spec.getId(), false);
+
     return createAndStartSupervisorInternal(spec, true);
   }
 
-  public void stopAndRemoveSupervisor(String id, boolean writeTombstone)
+  public synchronized boolean stopAndRemoveSupervisor(String id)
   {
-    Pair<Supervisor, SupervisorSpec> pair = supervisors.get(id);
-    if (pair != null) {
-      if (writeTombstone) {
-        metadataSupervisorManager.insert(id, new NoopSupervisorSpec()); // where NoopSupervisorSpec is a tombstone
-      }
-      pair.lhs.stop(true);
-      supervisors.remove(id);
-    }
+    Preconditions.checkNotNull(id, "id");
+
+    return possiblyStopAndRemoveSupervisorInternal(id, true);
   }
 
   @LifecycleStart
-  public void start()
+  public synchronized void start()
```

Comment:
I would be ok with using synchronized(this) in this case, can't see any advantage of using a new lock object. In general, no code should be using instance of SupervisorManager etc as a lock and using "this" should be ok then.